src/client/client_impl.cc

```diff
@@ -288,21 +320,256 @@ google::protobuf::util::StatusOr<std::pair<size_t, void*>> ClientImpl::ReadFile(
+                                           primary_server_address));
+    // Issue WriteFileChunkRequest and check status
+    StatusOr<WriteFileChunkReply> write_reply_or(
+        primary_server_service_client->SendRequest(write_request, client_context));
+
+    // Handle grpc error, and retry if possible
+    if (!write_reply_or.ok()) {
+      LOG(ERROR) << "Send write chunk data request to " 
+                 << primary_server_address << "failed due to " 
+                 << write_reply_or.status();
+    } else {
+      auto write_reply(write_reply_or.ValueOrDie());
+      switch (write_reply.status()) {
+        case FileChunkMutationStatus::OK:
+          LOG(INFO) << "Write to file " << filename << " at chunk_index "
+                    << chunk_index << " and offset " << offset << " for "
+                    << nbytes << " bytes succeeds";
+          return write_reply_or;
+        case FileChunkMutationStatus::FAILED_NOT_LEASE_HOLDER:
+          // Client found out that the primary it though is no longer a lease
+          // holder, we need to force to get the chunk metadata and retry
+          LOG(INFO) << "Retrying as primary server "
+                    << primary_server_address << " is no longer lease holder "
+                    << "refetching chunk metadata";
+          get_metadata_status = 
+              GetMetadataForChunk(filename, chunk_index, 
+                                  OpenFileRequest::WRITE, chunk_handle, 
+                                  chunk_version,
+                                  chunk_server_location_entry, true);
+          if (!get_metadata_status.ok()) {
+            return get_metadata_status; 
+          }
+          break;
+        default:
+          // For other cases, we consider them to be irrecoverable and return
+          // error
+          LOG(ERROR) << "Write to file " << filename << " at chunk_index "
+                     << chunk_index << " and offset " << offset << " for "
+                     << nbytes << " byte failed due to " 
+                     << write_reply.status();
+          return write_reply_or;
+      }
+    }
+
+    retry--;
+  }
+
+  LOG(ERROR) << "Write file chunk failed after 3 retries";
+  return google::protobuf::util::Status(
+             google::protobuf::util::error::INTERNAL,
+                 "Write file chunk failed after 3 retries"); 
+}
+
+google::protobuf::util::Status ClientImpl::WriteFile(const char* filename, 
+    void* buffer, size_t offset, size_t nbytes) {
+  const size_t chunk_block_size(config_manager_->GetFileChunkBlockSize() *
+                                common::bytesPerMb);
+  // Record the number of bytes that we need to write
+  size_t bytes_written(0);
+  // Record the number of bytes to write
+  size_t remain_bytes(nbytes);
+  // A flag to detect errors that encountered during issuing writes
+  size_t chunk_start_offset(offset % chunk_block_size);
+
+  for (size_t chunk_index = offset / chunk_block_size; remain_bytes > 0; 
```

Comment:
should we parallelize this?

Reply:
Yes, let's do so after we have a stable end-to-end version if time allows?